File: Algorithms/py/067.AddBinary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        if len(a) < len(b):
            a, b = b, a
        m, n, p, s = len(a), len(b), 0, ""
        i = 0
        while i < min(m, n):
            q = ord(a[m-1-i]) - ord('0') + ord(b[n-1-i]) - ord('0') + p
            p = q // 2
            s += str(q%2)
            i += 1
        while i < m:
            q = ord(a[m-1-i]) - ord('0') + p
            p = q // 2
            s += str(q%2)
            i += 1
        if p != 0:
            s += '1'
        return s[::-1]

File: Algorithms/cpp/067.AddBinary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        int m = a.size();
        int n = b.size();
        int p = 0, i;
        string s;
        for (i=0; i < m && i < n; i++) {
            int q = a[m-1-i] - '0' + b[n-1-i] - '0' + p;
            p = q / 2;
            if (q%2 == 1) {
                s.insert(0, 1, '1');
            }
            else {
                s.insert(0, 1, '0');
            }
        }
        for (; i < m; i++) {
            int q = a[m-1-i] - '0' + p;
            p = q / 2;
            if (q%2 == 1) {
                s.insert(0, 1, '1');
            }
            else {
                s.insert(0, 1, '0');
            }
        }
        for (; i < n; i++) {
            int q = b[n-1-i] - '0' + p;
            p = q / 2;
            if (q%2 == 1) {
                s.insert(0, 1, '1');
            }
            else {
                s.insert(0, 1, '0');
            }
        }
        if (p != 0) {
            s.insert(0, 1, '1');
        }
        return s;
    }
};
